### DISCUSSION TITLE
WIP - Nuxt v3 SEO Opportunities

### DISCUSSION THREAD

**Discussion by harlan-zw** [nuxt/nuxt#17658] [Ideas]
## Introduction

Building a search engine friendly website will always be a moving target. Ignoring the content opportunities, there's a set of technical requirements that any site that wishes to have the best organic growth potential needs to meet.

These technical requirements can be summariesed as:
- Compliant HTML (meta tags, img meta, img sizes, header tag hierarchy etc)
- Crawling (sitemap, robots.txt, internal links, canonical links, redirects)
- Accessibility (mobile-friendly, misc link/images issues)
- Performance ([web vitals](https://web.dev/vitals/))
- SEO Sugar - optional (AMP, Schema.org, share images)

By looking at the state of Nuxt SEO, we can identify opportunities for improvement and think about what a Nuxt SEO Module might look like.

## State of Nuxt v3 SEO

In Nuxt v2 we had a solid ecosystem of modules to choose from.  For nuxt v3 things aren't looking as good (yet):

**Compliant HTML**
- [Meta SEO](https://github.com/nuxt/framework/discussions/1823) 🔨 - In Development
- [nuxt/html-validator](https://html-validator.nuxtjs.org/) ⌛ - not blocked: can be worked on 
- [nuxt/i18n](https://i18n.nuxtjs.org/) - optional 🔨 - In Development

**Crawling**
- [nuxt/sitemap](https://sitemap.nuxtjs.org/) ⌛ - blocked: waiting for nitropack / static generation
- [nuxt/robots](https://github.com/nuxt-community/robots-module) ⌛ - blocked: waiting for nitropack / static generation
- [nuxt/redirects](https://github.com/nuxt-community/redirect-module) 💀 - deprecated: redirects will be a core feature

**Accessibility**
- [@unlighthouse/nuxt](https://unlighthouse.dev) ✅

**Performance**
- [nuxt/web-vitals](https://github.com/nuxt-community/web-vitals-module) ⌛ - not blocked: can be worked on 
- [nuxt/image](https://image.nuxtjs.org/) 🔨 - In Development
- nuxt/script 🔨 - In Development
- [nuxt/partytown](https://github.com/nuxt-community/partytown-module) ✅
- [nuxt-delay-hydration](https://github.com/harlan-zw/nuxt-delay-hydration) ⌛ - not blocked: can be worked on 

**SEO Sugar - optional**
- [nuxt/amp](https://github.com/nuxt-community/amp-module) - [no longer relevant for SEO](https://plausible.io/blog/google-amp)
- [nuxt-jsonld](https://github.com/ymmooot/nuxt-jsonld) ⌛ - not blocked: can be worked on 

## Core Opportunities

With the migration of these Nuxt v2 packages to Nuxt v3, it would be ideal to look for opportunities where module boilerplate can be reduced through core Nuxt additions.

### Site Meta Config

A core way to define global site meta details, such as the hostname. 

```ts
siteMeta: {
  // needed for sitemap, canonical and og links
  hostname: 'example.com',
}
```

Site meta could be merged into page meta with a configurable merge function, giving access siteMeta access to `useMeta`.

Outcomes:
- boilerplate removed from Meta SEO, sitemap, robots.txt

Concerns:
- Extra configuration key, could use existing env or runtime config
- Could be simplified to just `hostname: string`

### Index Opt-Out

Introduce a standard way developers can mark pages as not-indexable. 

```ts
definePageMeta({
  indexable: false // or `noindex: true`
})
```

Disabling indexing site wide
```ts
siteMeta: {
  indexable: false, // or `noindex: true`
}
```

Outcomes:
- Sitemap generation for static pages can be automated
- Robots.txt can be automatically generated

### Handling Dynamic Route URLs

In nuxt v2 for SSG's, there's duplication on providing the list of dynamic route URLs for sitemaps and the generate function. Ideally, there'd be some core way to discover all URLs that the final site will be using.

I have two ideas on how it could work:

1. Devs provide a function within the page component on how all routes for that file will be resolved.

```ts
<script setup lang="ts">
import { fetchPost, fetchPosts } from '@/data'

defineDynamicRoutes(async () => {
  return (await fetchPosts().map(p => {
    return {
      params: {
        slug: p.slug
      },
      payload: p
    }
  }))
})

const post = useAsync(() => fetchPost({ slug: useRoute().params.slug }))
</script>
```

Outcomes:
- Devs won't need to compute the final URL of the page, it can be computed from the params and file
- Logic for data fetching a single and multiple of the page payload is togethor

Concerns:
- Implementation isn't exactly clear, will need some thought

2. Leverage siteMeta to define all routes

An alternative option would be to use the above-mentioned site meta config key to define all routes. This data would be merged with any meta defined on the page component itself.

```ts
siteMeta() {
  const pages = await fetchPages().map(p => normaliseForRoute(p))
  return {
    // needed for sitemap, canonical and og links
    hostname: 'example.com',
    routes: [
      { path: '/', pageMeta: { title: 'test', indexable: false } }
      ...pages
    ]
  }
}
```

Outcomes:
- Managing page meta for an entire site becomes easier, as you'd be able to see all of it in one spot
- Sometimes you'll need a list of all pages at runtime, such as generating a dynamic nav list. We could expose the site meta, while stripping payloads for this use case. Similar to how the current routes.json is used in Nuxt v2 to figure out if a link is a 404. 
- Implementation is fairly straightforward

Concerns:
- Final paths need to be computed
- Loading in an extra file to discover routes is potentially expensive for large sites, build time would be better 

### Module: Nuxt Crawl

Automatically generates sitemap.xml and robots.txt based on the configuration of the above. Can be extended to build these files however you want.

### Package: VueUse useSchema

-----------------

Still working on this - WIP

### Nuxt/image

- better support for generating image sizes fitting to a specific grid system

# Nuxt SEO Module

- wrapper for a bunch of nuxt v3 SEO modules

**Comment by pi0** [comment]
Thanks so much for well-written ideas @harlan-zw 💚 

**A quick state about mentioned modules:**

State of `sitemap`, `robots` and `image` modules are still pending for nitropack and static generation support. Once landed, I believe we could directly migrate them, and `redirect` module will be directly a feature of nuxt route rules.

Support of `html-validator`, `web-vitals`, `jsonld` and amp ([however is not relevant to SEO anymore](https://plausible.io/blog/google-amp)) and `nuxt-delay-hydration` (?) i believe is not blocked by anything with the current state of Nuxt 3 and matter of time only.

**hostname and canonical url:**

I think it this absolutely a great idea to have a built-in that resolves to the active or canonical hostname. We could use [app namespace](https://github.com/nuxt/framework/blob/main/packages/schema/src/config/_app.ts#L33) which is a **runtime config** to introduce a new option (`hostname`|`domain`|`canonicalHost`), a built-in composable like `getURL('/foo')` that resolves a path to the full URL with hostname and add canonical meta if app option is provided.

Would be a nice idea if at the same time we integrate a composable with [unjs/requrl](https://github.com/unjs/requrl) that gives _current hostname_ in order to generate URLs with the active hostname and use it as a fallback when `canonicalDomain` is not explicitly set to resolve URLs.

(will add follow-up issue to escalate this)

**Dynamic urls and paths:**

I like the idea of `defineDynamicRoutes`. As you can see, implementation could be tricky since unlike macros like `definePage` meta, we couldn't just extract it as it could depend on any composable or vue app runtime and besides, we need to find a way to communicate with a nitro builder in order to extend routes but I'm pretty sure we could find a way to make this idea happen ;)

Regarding the simpler method you mentioned as with `siteMeta` returning routes, Nitropack, will provide a `routes` config and possibly a built-in crawler. We could reuse the same infrastructure in Nuxt 3 for sharing logic between SEO modules that require such a list of all available dynamic routes or possibly also give them to the runtime for navigation purposes, etc.

The only tricky note is that with Nuxt3, unlike Nuxt2 not all routes are generated at built-time with progressive static generation. Therefore it is not always possible (without cost of one time full crawl) to predict all dynamic routes but only static generated dynamic routes.

This is a rather wide topic to discuss and beyond only SEO requirements. I suggest continuing the seo-aspect of dynamic URLs usage once basic functionality landed.

**Other mentioned modules:**

- Nuxt crawl: Same as Nuxt 2 it will be a part of Nitro with even better possibilities to crawl api routes. I Belive we could keep depending on community modules to provide sitemap generation functionality since it is not always as simple as a 1-1 mapping between all routes to sitemap nor could be a generic enough functionality being part of the core.
- Nuxt image: A discussion in [nuxt/image](https://github.com/nuxt/image) is more than welcome for imrove
- Nuxt SEO module: SEO is rather a wide topic. We are going to bring as much as possible of SEO-related best practices to the core (https://github.com/nuxt/framework/discussions/1823). We could instead introduce a collective guide in the concepts section of Nuxt 3 that advises best practices with built-in SEO related features as well as listing modules that could benefit website's SEO to opt-in such as sitemap once ready.

**Reply by harlan-zw** [reply]
Thank you very much for the detailed response @pi0, I will revise this accordingly over the next few days.

**Comment by gustavorps** [comment]
Current static build don't generate pages with header :-[

`page.index.vue`
```vue
<template>
  <div>Hello World</div>
</template>
<script>
export default {
  data: () => ({
    description: 'Page description',
  }),
  head() {
    return {
      title: 'Page title',
      meta: [
        {
          hid: 'description',
          name: 'description',
          content: this.description
        },
        {
          hid: 'og:description',
          name: 'og:description',
          content: this.description
        },
        {
          hid: 'og:image:secure_url',
          name: 'og:image:secure_url',
          content: 'https://via.placeholder.com/600x400'
        },
        {
          hid: 'og:type',
          name: 'og:image:secure_url',
          content: 'website'
        },
      ]
    }
  },
  props: {
    name: String,
    description: String,
    image: Object,
  }
}
</script>
```
`dist/index.html`
```html
<!DOCTYPE html>
<html >

<head >
  <link rel="stylesheet" href="/_nuxt/entry.0d7db7e1.css">
</head>

<body >
  <div id="__nuxt">...
</body>

</html>
```
